panther_lights/src/animation/animation.py

```diff
@@ -34,11 +38,40 @@ def __init__(self, animation_description: dict, num_led: int, controller_freq: f
                 raise KeyError('Brightness has to match boundaries 0 < brightness <= 1')
             self._brightness = int(round(self._brightness * 100))
 
-    def __call__(self) -> None:
-        '''returns new frame'''
-        raise NotImplementedError
+        # evaluate number of animation frames
+        self._anim_len = int(round(self._duration * controller_freq))
+        self._full_anim_len = self._anim_len * self._loops
+
+        if self._anim_len < 1:
+            raise KeyError('Animation duration is too short to display with the current frequency')
+
+    def __call__(self) -> Generator:
+        if self._current_cycle <= self._loops:
+            frame = self._update_frame()
+            self._anim_iteration += 1
+            self._progress = (
+                self._anim_iteration + self._anim_len * (self._current_cycle - 1)
+            ) / self._full_anim_len
+
+            if self._anim_iteration >= self._anim_len:
+                self._anim_iteration = 0
+                self._current_cycle += 1
+
+            if self._current_cycle > self._loops:
+                self._finished = True
+
+            for led in frame:
+                yield led
```

Comment:
```suggestion
            frame = self._update_frame()
            self._anim_iteration += 1
            self._progress = (
                self._anim_iteration + self._anim_len * (self._current_cycle - 1)
            ) / self._full_anim_len

            if self._anim_iteration > self._anim_len:
                self._anim_iteration = 0
                self._current_cycle += 1

            if self._current_cycle > self._loops:
                self._finished = True
                
            return frame
        return [0] * self._num_led
```

After conversation, let's give up the idea of generators and return empty frame in case of end of animation.